### PR TITLE
Fix cancel issue between Query Frontend and Query Schdeduler

### LINKED ISSUE #5132: Request `cancel` issue between loki `QueryFrontend` and `QueryFrontendWorker`
# LogQL request cancellation not propogated.

## Problem.
When LogQL request is cancelled by the client(LogCLI or Grafana), the cancellation is received by Query Frontend, but not propagated to some of the downstream requests started by Query Frontend(eg to queriers mainly).

This leads to more resource conception in Loki because, even after original request is cancelled, the queriers started by the request is still running.

NOTE: Can happen for any LogQL queries

## Steps to reproduce the issue.

NOTE: Here I use LogCLI, for two reasons

1. Grafana sends two queries (actual log query and log volume query if enabled) so narrowing down the resource consumption is bit hard during that time interval. Whereas, using LogCLI we can send only single metric query that is causing the issue.
1. Easy to change the request timeout value for investigation.

### 1. Make `count_over_time` query via LogCLI
We use `timeout 10s` that closes the client connection after 10s _cancelling_ the request.

Here we use `date | md5sum | cut -d\- -f1` for random ID in the query (easy to find the exact query).

```
QID=$(date | md5sum | cut -d\- -f1) && \
	echo $QID && \
	timeout 10s \
	logcli query 'sum by (level) (count_over_time({cluster="<cluster-name>"} |= "'"$QID"'"[1m]))'\
	--since=48h \
	-q
```

Above command will print random id. Copy that to search for that query to find it's traces.

### 2. Search for that query in Loki.
Go search for the query we made on Grafana Explore or LogCLI via following query.
fill-in `<random-id>` copied from the previous step.

```
{cluster="<cluster>", namespace="<namespace>", job="<namespace>/query-frontend"} |= "caller=metrics.go" |= "<random-id>"
```

### 3. Tempo traces.
Optionally see the traces of the requests.

tempo-traces

You can see it takes more than 10s (something was running after the cancelation)

Also from the traces, you can see, Query Frontend got the cancellation and returned 499 status correctly (in 10s)

So the problem is **even Query Frontend got the cancellation and responded correctly with 499 at 10s, the complete request cycle ran for more than 15s (this went upto 45s in some cluster depends on the traffic)**

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
Query Frontend failed to send cancel signal even to Query Frontend Worker during some special cases(explained below)
And that makes querier to run till completion(even after cancellation) wasting lots of resources.

This scenario is [analogous to this simple Go program](https://go.dev/play/p/YSb64s9UF6u) (run in go-playground). The gist is `default` branch in go `select` loop takes always precedence when other branch operations are blocked (sending or receiving on the channel). In our case, we **missed the sending of cancel signal**

Now to the original issue.

The main observation is for every Querier that haven't received cancel, it's corresponding Scheduler also didn't get it.

1. single query on QF has split into 128 sub queries
2. And all of them got scheduled and picked up by the querier.
3. I added some logs in two points. Where QF send sending cancel to scheduler, And scheduler loop sends it to querier. 

Then among those 128 subqueries, only 8 of them got cancel from QF to Scheduler, remaining 120 got dropped in QF itself, never even reached the Scheduler.

The root cause is, whenever frontend worker loop is busy with service request on the main select branch ([ lisitening on w.requestCh](https://github.com/grafana/loki/blob/main/pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go#L278-L334) ), and if during that time, QF [send cancel request to Scheduler via cancelCh](https://github.com/grafana/loki/blob/main/pkg/lokifrontend/frontend/v2/frontend.go#L229-L230) it will get block for sometime on the other end (coz no buffer).  **During that time, it got into default case and thus ignoring the cancel signal even before sending it to Scheduler**.

Basically we need to make sure to send cancel signal to Scheduler, no matter how busy the frontend worker to pickup cancel from the cancelCh.

**Which issue(s) this PR fixes**:
Fixes #5132 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go [hunk 1]
```diff
@@ -325,6 +327,7 @@ func (w *frontendSchedulerWorker) schedulerLoop(loop schedulerpb.SchedulerForFro
 			}
 
 		case reqID := <-w.cancelCh:
+			fmt.Println("Got the cancel as well", reqID)
```
Comment (kavirajk) [inline pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:330]: Still keeping the comments, in case reviewer wants to run the test with and without fix and easy to understand where cancel is getting lost.

Will do cleanup after initial review

File: pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go [hunk 2]
```diff
@@ -146,6 +147,20 @@ func (f *frontendSchedulerWorkers) getWorkersCount() int {
 	return len(f.workers)
 }
 
+// sendRequestCancel sends cancellation to the "already scheduled" frontendRequest.
+// It will make sure the frontend worker that is responsible for the `reqID`
+// receives the cancel signal.
+func (f *frontendSchedulerWorkers) sendRequestCancel(reqID uint64, cancelCh chan<- uint64) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
```
Comment (cyriltovena) [inline pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:155]: is the lock just for f.workers ? if yes please then test it and release the lock before blocking on the send of the channel.

Comment (kavirajk) [reply pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:155]: Yea. in fact that was my initial fix using already existing [f.workers.getWorkersCount](https://github.com/grafana/loki/blob/3e3e23826b8ab1b149faa04387fce90ebdc79b47/pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go#L143). Just get `len(f.workers)` with locks and do the channel operation after that. But that may lead to a race right?

In fact it may leads to two problems.

say If I do like this (`getWorkersCount()` release lock already).

````
if f.workers.getWokersCount() > 0 {
            cancelCh <- f.requestID
}
````

1. There is a race between `if getWorkersCount() > 0` and `cancelCh <- f.requestID`.
2. If workers count > 0 during if, and it goes to 0 during `cancelCh <- f.requestID` then it goes for waiting it forever, because no one is listening on the other side of the channel (because no frontend workers)

It cannot go to blocking on send if we have non-zero frontend worker running. And both non-zero count check and send should happen with lock.

Comment (cyriltovena) [reply pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:155]: Yeah that seems correct you are right !

Comment (cyriltovena) [reply pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:154]: Also for other reviewer the lock is only held when adding/starting or removing workers. That should be fast and non blocking.

File: pkg/lokifrontend/frontend/v2/frontend_test.go [hunk 3]
```diff
@@ -176,6 +176,45 @@ func TestFrontendCancellation(t *testing.T) {
 	})
 }
 
+// Bug: If FrontendWorkers are busy, cancellation passed by Query frontend may not reach
+// all the frontend workers. (assumming we run multiple frontend workers)
+func TestFrontendWorkerCancellation(t *testing.T) {
```
Comment (cyriltovena) [inline pkg/lokifrontend/frontend/v2/frontend_test.go:181]: Run the test and thanks it does cover the case.

File: pkg/lokifrontend/frontend/v2/frontend.go [hunk 4]
```diff
@@ -223,12 +224,9 @@ enqueueAgain:
 	select {
 	case <-ctx.Done():
 		if cancelCh != nil {
-			select {
-			case cancelCh <- freq.queryID:
-				// cancellation sent.
-			default:
-				// failed to cancel, ignore.
-			}
+			// NOTE(kavi): I think we don't need buffer channel.
```
Comment (cyriltovena) [inline pkg/lokifrontend/frontend/v2/frontend.go:227]: remove the note.

File: pkg/lokifrontend/frontend/v2/frontend.go [hunk 5]
```diff
@@ -204,6 +204,7 @@ enqueueAgain:
 
 	select {
 	case <-ctx.Done():
+		level.Warn(f.log).Log("msg", "request enqueued, but not cancelable", "queryID", freq.queryID)
```
Comment (cyriltovena) [inline pkg/lokifrontend/frontend/v2/frontend.go:207]: I requested this because I have doubt this is practical and if it is we're enqueuing without being able to cancel which can lead to the same bug as before.

**Review by cyriltovena — APPROVED** [review]
LGTM

@colega PTAL I want to make sure we didn't miss anything.

File: pkg/lokifrontend/frontend/v2/frontend_test.go [hunk 6]
```diff
@@ -176,6 +176,45 @@ func TestFrontendCancellation(t *testing.T) {
 	})
 }
 
+// Bug: If FrontendWorkers are busy, cancellation passed by Query frontend may not reach
+// all the frontend workers thus not reaching the scheduler as well.
+func TestFrontendWorkerCancellation(t *testing.T) {
+	f, ms := setupFrontend(t, nil)
+
+	// fmt.Println("workers count", f.schedulerWorkers.getWorkersCount(), "max-concurrency per worker", f.cfg.WorkerConcurrency)
+
```
Comment (colega) [inline pkg/lokifrontend/frontend/v2/frontend_test.go:185]:
```suggestion
	// fmt.Println("workers count", f.schedulerWorkers.getWorkersCount(), "max-concurrency per worker", f.cfg.WorkerConcurrency)

```

**Comment by colega** [comment]
I agree that this fixes the issue but I think we're overcomplicating it a little bit. Cancelling the querier request isn't a strict need but just politeness and avoids wasting resources, so _it would be great to ensure it, but not application-critical_.

If there's no need to ensure that cancellation has happened, I think just adding some sensible capacity to each worker's `cancelCh` should be enough (while keeping the `default: pass` section, to avoid blocking if that cancellation channel is really full).

Let's say `make(chan uint64, 1024)`: if we have more than 1024 canceled requests waiting for worker to be free, then we probably have bigger issues to worry about anyway.

WDYT?

**Comment by cyriltovena** [comment]
> I agree that this fixes the issue but I think we're overcomplicating it a little bit. Cancelling the querier request isn't a strict need but just politeness and avoids wasting resources, so _it would be great to ensure it, but not application-critical_.
> 
> If there's no need to ensure that cancellation has happened, I think just adding some sensible capacity to each worker's `cancelCh` should be enough (while keeping the `default: pass` section, to avoid blocking if that cancellation channel is really full).
> 
> Let's say `make(chan uint64, 1024)`: if we have more than 1024 canceled requests waiting for worker to be free, then we probably have bigger issues to worry about anyway.
> 
> WDYT?
> 
> Edit, I tried with this:
> 
> ```diff
> diff --git pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go
> index fa6395783..7d32f2f70 100644
> --- pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go
> +++ pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go
> @@ -191,7 +191,7 @@ func newFrontendSchedulerWorker(conn *grpc.ClientConn, schedulerAddr string, fro
>                 schedulerAddr: schedulerAddr,
>                 frontendAddr:  frontendAddr,
>                 requestCh:     requestCh,
> -               cancelCh:      make(chan uint64),
> +               cancelCh:      make(chan uint64, 1024),
>         }
>         w.ctx, w.cancel = context.WithCancel(context.Background())
> ```
> 
> And the test from this PR also passes 👍

Yeah this interesting 🤔  + the lock could be expensive.

File: pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go:154]: That seems correct yep.

Comment (kavirajk) [reply pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go]: @colega you are absolutely right, `cancelCh` that is part of frontendReq belongs to single woker. And hard to assume if that worker will ever get that.

Smart thinking :)

**Review by colega — APPROVED** [review]
